File: jpa-box-colors/src/test/java/com/yonlabs/java_boxcolors/owning/OwningBoxColorTest.java
package com.yonlabs.java_boxcolors.owning;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import javax.persistence.EntityManager;
import javax.transaction.Transactional;
import java.util.Arrays;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
@Transactional
public class OwningBoxColorTest {

    @Autowired
    EntityManager entityManager;

    @Test
    public void testGetBox_ColorSaved() {
        // given: create a new box with one color
        int boxId = 1;
        int colorId = 1;
        JColor1 color = new JColor1(new JColorId1(colorId));
        JBox1 box = new JBox1(boxId, Arrays.asList(color));
        entityManager.persist(box);
        // forces flush to a database
        entityManager.flush();
        // detaches all managed entities to force find from a database
        entityManager.clear();

        // when: find a box
        JBox1 found = entityManager.find(JBox1.class, boxId);

        // then:
        assertThat(found).isNotNull();
        assertThat(found.getColors().size()).isEqualTo(1);
    }

}

File: jpa-box-colors/src/test/kotlin/com/yonlabs/kt_boxcolors/OwningBoxColorTest.kt
package com.yonlabs.kt_boxcolors

import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import javax.persistence.EntityManager
import javax.transaction.Transactional

@SpringBootTest
@Transactional
open class OwningBoxColorTest @Autowired constructor(
        val entityManager: EntityManager
) {
    @Test
    fun testGetBox_ColorSaved() {
        // given: create a new box with one color
        val boxId = 1
        val colorId = 1
        val color = Color1(ColorId1(colorId))
        val box = Box1(boxId, mutableListOf(color))

        entityManager.persist(box)
        // forces flush to a database
        entityManager.flush()
        // detaches all managed entities to force find from a database
        entityManager.clear()

        // when: find a box
        val found = entityManager.find(Box1::class.java, boxId)

        // then:
        assert(found != null)
        assert(found?.colors?.size == 1)
    }

}